rts/Game/UI/SelectionKeyHandler.cpp

```diff
@@ -124,6 +134,12 @@ namespace {
 	DECLARE_FILTER(Idle, unit->commandAI->commandQue.empty())
 	DECLARE_FILTER(Waiting, !unit->commandAI->commandQue.empty() && (unit->commandAI->commandQue.front().GetID() == CMD_WAIT))
 	DECLARE_FILTER(Guarding, !unit->commandAI->commandQue.empty() && (unit->commandAI->commandQue.front().GetID() == CMD_GUARD))
+  DECLARE_FILTER(Patrolling, !unit->commandAI->commandQue.empty() && (
```

Comment:
Happy to have anything better than this here

Reply:
1) could document why 4.
2) there could be a nice `CommandAI::isPatrolling` helper function and then you'd just do `unit->commandAI->isPatrolling()` in this macro.
3) when extracted to a function it could become a loop. `for (int i = 0; i < 3; ++i) if (size >= i && at(i) == patrol)`

Reply:
1. Good call, in most scenarios a unit patrolling will be either in a pure patrol (1), move+patrol(2), engaged in something (reclaim/build/fight something on the way) OR 2 moves + patrol (4). I found 4 to be the sweet spot a player would like to engage the filter
2. I was a bit hesitant in doing that but after this monstruosity I created I think it has its merits, but we want it to specify a cmd depth to avoid problematic cases (-1 full queue)

Reply:
> 1. Good call, in most scenarios a unit patrolling will be either in a pure patrol (1), move+patrol(2), engaged in something (reclaim/build/fight something on the way) OR few moves + patrol (4). I found 4 to be the sweet spot a player would like to engage the filter

Do I understand correctly that a unit's `queue.front()` action would be, for example, `CMD_FIGHT` if the unit is patrolling but currently fighting another unit? And similarly, if a unit is moving to the position that the user has requested them to patrol, their `queue.front()` might be `CMD_MOVE`?

Reply:
Patrol works by prepending Fight commands alongside the edges of the patrol route (Fight is what other games call Attack-Move). So if nothing is going on, a patrolling unit's first command will be Fight and 2nd+ will be Patrol.
![image](https://github.com/beyond-all-reason/spring/assets/2573076/9e776547-ad58-4e45-a6c3-d188ca273b35)

Fight works by prepending Attack commands when there's a target. It also makes sure to attack-move back to the point whence it originally engaged the enemy, before following back on that course. In that case the Patrol command can be 4th in the queue.
![image](https://github.com/beyond-all-reason/spring/assets/2573076/5c06d1c4-78db-46b1-848a-4286f55d93f6)

One could imagine more commands that would push the Patrol command even further back. For example tactical AI that prepends Move commands in front of Attack to dodge shots, or a "set speed" decorator command in front of the movement-related commands that makes sure a patrolling group made of different units sticks together.

Reply:
Just read your comment in the other PR

> the queue gets prepended by other commands as the unit does stuff

So I think I am following correctly. If this is the cas,e, then is there a reason this logic does not exist for the guarding filter? Shouldn't the same logic (that a unit might be moving before guarding something) apply?

Another implementation option: add a function like `contains(command, depth)` to the `CCommandQueue` class itself. Might be overkill if this kind of queue inspection is only valuable here, though.

Reply:
The guarding filter works because the guard command doesn't prepend other orders, it handles everything on its own.